File: ai/sdk/yuga.py
"""
Yuga Python SDK
pip install requests  (only dependency)

Usage:
    from yuga import Yuga

    yuga = Yuga(api_key='yuga_live_xxx', api_url='https://yoursite.com/yuga/api/')
    reply = yuga.chat('What is your pricing?')
    print(reply['reply'])

    # Multi-turn conversation (memory is automatic)
    r1 = yuga.chat('What is the pro plan?')
    r2 = yuga.chat('How much does it cost?')   # remembers context

    # Text generation (real LM)
    text = yuga.generate('Our pricing starts at', max_chars=200, temperature=0.8)

    # Train on your platform
    yuga.learn_site('https://yourplatform.com', max_pages=30)
"""

import json
import time
from typing import Generator, Optional

try:
    import requests as _requests
    _has_requests = True
except ImportError:
    import urllib.request
    import urllib.error
    _has_requests = False


class YugaError(Exception):
    pass


class Yuga:

    def __init__(
        self,
        api_key: str = '',
        api_url: str = '',
        model:   str = 'default',
        timeout: int = 30,
    ):
        self.api_key = api_key
        self.api_url = api_url.rstrip('/') + '/'
        self.model   = model
        self.timeout = timeout
        self._history = []  # conversation memory

    # ── Core request ─────────────────────────────────────────────────
    def _req(self, action: str, body: dict = None, method: str = 'POST') -> dict:
        url     = self.api_url + '?action=' + action
        headers = {'Content-Type': 'application/json', 'Accept': 'application/json'}
        if self.api_key:
            headers['X-API-Key'] = self.api_key

        payload = body or {}
        payload.setdefault('model', self.model)
        data = json.dumps(payload).encode()

        if _has_requests:
            r = _requests.request(method, url, headers=headers, data=data, timeout=self.timeout)
            result = r.json()
        else:
            req = urllib.request.Request(url, data=data, headers=headers, method=method)
            try:
                with urllib.request.urlopen(req, timeout=self.timeout) as resp:
                    result = json.loads(resp.read())
            except urllib.error.HTTPError as e:
                result = json.loads(e.read())

        if not result.get('ok') and result.get('error'):
            raise YugaError(result['error'])
        return result

    # ── Chat ─────────────────────────────────────────────────────────
    def chat(self, message: str, model: str = None) -> dict:
        """Send a message. Conversation memory is automatic."""
        r = self._req('brain_chat', {
            'message': message,
            'history': self._history,
            'model':   model or self.model,
        })
        if 'history' in r:
            self._history = r['history']
        return r

    # ── Generate ─────────────────────────────────────────────────────
    def generate(
        self,
        prompt:      str,
        max_chars:   int   = 200,
        temperature: float = 0.8,
        top_p:       float = 0.9,
        model:       str   = None,
    ) -> dict:
        """Generate text using the trained YugaGen model."""
        return self._req('yugagen_chat', {
            'prompt':      prompt,
            'max_chars':   max_chars,
            'temperature': temperature,
            'top_p':       top_p,
            'model':       model or self.model,
        })

    # ── Learn ─────────────────────────────────────────────────────────
    def learn_text(self, text: str, source: str = 'sdk') -> dict:
        """Train on raw text."""
        return self._req('learn_text', {'text': text, 'source': source})

    def learn_url(self, url: str) -> dict:
        """Fetch and train on a single URL."""
        return self._req('learn_url', {'url': url})

    def learn_site(self, url: str, max_pages: int = 20) -> dict:
        """Crawl a full website and train on all pages."""
        return self._req('learn_site', {'url': url, 'max_pages': max_pages})

    # ── Status ────────────────────────────────────────────────────────
    def status(self, model: str = None) -> dict:
        """Get model training status."""
        return self._req('status', {'model': model or self.model})

    def models(self) -> list:
        """List all available models."""
        r = self._req('models', {})
        return r.get('models', [])

    def yugagen_status(self, model: str = None) -> dict:
        """Get YugaGen checkpoint status."""
        return self._req('yugagen_status', {'model': model or self.model})

    # ── Feedback ──────────────────────────────────────────────────────
    def feedback(self, question: str, answer: str, score: int, session_id: str = '') -> dict:
        """Submit thumbs up (1) or thumbs down (-1) on an answer."""
        return self._req('feedback', {
            'question':   question,
            'answer':     answer,
            'score':      score,
            'session_id': session_id,
        })

    # ── Memory ────────────────────────────────────────────────────────
    def reset_memory(self):
        """Clear conversation history."""
        self._history = []

    def get_history(self) -> list:
        """Get current conversation history."""
        return list(self._history)

    # ── Async batch ───────────────────────────────────────────────────
    def batch_learn(self, texts: list, delay: float = 0.2) -> list:
        """Train on multiple texts with a delay between each."""
        results = []
        for text in texts:
            results.append(self.learn_text(text))
            time.sleep(delay)
        return results

    def __repr__(self):
        return f"Yuga(api_url='{self.api_url}', model='{self.model}')"


# ── Async variant (requires aiohttp) ─────────────────────────────────
try:
    import aiohttp
    import asyncio

    class AsyncYuga(Yuga):
        """Async version of the Yuga SDK. Requires aiohttp."""

        async def _req_async(self, action: str, body: dict = None) -> dict:
            url     = self.api_url + '?action=' + action
            headers = {'Content-Type': 'application/json'}
            if self.api_key:
                headers['X-API-Key'] = self.api_key
            payload = body or {}
            payload.setdefault('model', self.model)

            async with aiohttp.ClientSession() as sess:
                async with sess.post(url, json=payload, headers=headers, timeout=aiohttp.ClientTimeout(total=self.timeout)) as resp:
                    result = await resp.json()
            if not result.get('ok') and result.get('error'):
                raise YugaError(result['error'])
            return result

        async def chat(self, message: str, model: str = None) -> dict:
            r = await self._req_async('brain_chat', {'message': message, 'history': self._history, 'model': model or self.model})
            if 'history' in r:
                self._history = r['history']
            return r

        async def generate(self, prompt: str, **kwargs) -> dict:
            return await self._req_async('yugagen_chat', {'prompt': prompt, **kwargs})

        async def learn_site(self, url: str, max_pages: int = 20) -> dict:
            return await self._req_async('learn_site', {'url': url, 'max_pages': max_pages})

except ImportError:
    pass  # aiohttp not available

File: ai/sdk/yuga.js
/**
 * Yuga JavaScript SDK
 * 
 * Install (copy to your project):
 *   <script src="https://yoursite.com/yuga/sdk/yuga.js"></script>
 *
 * Or use directly in Node.js:
 *   const Yuga = require('./yuga.js');
 *
 * Usage:
 *   const yuga = new Yuga({ apiKey: 'yuga_live_xxx', apiUrl: 'https://yoursite.com/yuga/api/' });
 *   const reply = await yuga.chat('What is your pricing?');
 *   console.log(reply.reply);
 */

(function (root, factory) {
  if (typeof module !== 'undefined' && module.exports) {
    module.exports = factory(); // Node.js
  } else {
    root.Yuga = factory(); // Browser
  }
}(typeof self !== 'undefined' ? self : this, function () {

  class Yuga {

    constructor(options = {}) {
      this.apiUrl  = (options.apiUrl  || options.api_url || '').replace(/\/?$/, '/');
      this.apiKey  = options.apiKey   || options.api_key  || '';
      this.model   = options.model    || 'default';
      this.timeout = options.timeout  || 30000;
      this._history = []; // conversation memory
    }

    // ── Core request ─────────────────────────────────────────────────
    async _request(action, body = {}, method = 'POST') {
      const url     = this.apiUrl + '?action=' + action;
      const headers = { 'Content-Type': 'application/json' };
      if (this.apiKey) headers['X-API-Key'] = this.apiKey;

      body.model = body.model || this.model;

      const ctrl = new AbortController();
      const timer = setTimeout(() => ctrl.abort(), this.timeout);

      try {
        const res  = await fetch(url, { method, headers, body: JSON.stringify(body), signal: ctrl.signal });
        const data = await res.json();
        clearTimeout(timer);
        if (!data.ok && data.error) throw new Error(data.error);
        return data;
      } catch (e) {
        clearTimeout(timer);
        throw e;
      }
    }

    // ── Chat (ChatGPT-style with memory) ─────────────────────────────
    async chat(message, options = {}) {
      const body = {
        message,
        history: this._history,
        model:   options.model || this.model,
      };
      const r = await this._request('brain_chat', body);
      if (r.history) this._history = r.history;
      return r;
    }

    // ── Generate (raw LM text generation) ─────────────────────────────
    async generate(prompt, options = {}) {
      return this._request('yugagen_chat', {
        prompt,
        max_chars:   options.maxChars   || 200,
        temperature: options.temperature || 0.8,
        top_p:       options.topP        || 0.9,
        model:       options.model       || this.model,
      });
    }

    // ── Learn from text ───────────────────────────────────────────────
    async learnText(text, source = 'sdk') {
      return this._request('learn_text', { text, source });
    }

    // ── Learn from URL ────────────────────────────────────────────────
    async learnUrl(url) {
      return this._request('learn_url', { url });
    }

    // ── Crawl entire site ─────────────────────────────────────────────
    async learnSite(url, maxPages = 20) {
      return this._request('learn_site', { url, max_pages: maxPages });
    }

    // ── Model status ──────────────────────────────────────────────────
    async status(model) {
      return this._request('status', { model: model || this.model });
    }

    // ── List models ───────────────────────────────────────────────────
    async models() {
      return this._request('models', {});
    }

    // ── Submit feedback (thumbs up/down) ──────────────────────────────
    async feedback(question, answer, score, sessionId = '') {
      return this._request('feedback', { question, answer, score, session_id: sessionId });
    }

    // ── Reset conversation memory ─────────────────────────────────────
    resetMemory() {
      this._history = [];
    }

    // ── Get conversation history ──────────────────────────────────────
    getHistory() {
      return [...this._history];
    }

    // ── Streaming chat (Server-Sent Events) ───────────────────────────
    async *stream(message, options = {}) {
      const url     = this.apiUrl + '?action=brain_chat_stream';
      const headers = { 'Content-Type': 'application/json' };
      if (this.apiKey) headers['X-API-Key'] = this.apiKey;

      const res = await fetch(url, {
        method: 'POST',
        headers,
        body: JSON.stringify({ message, model: options.model || this.model, stream: true }),
      });

      if (!res.ok) throw new Error('Stream request failed: ' + res.status);
      if (!res.body) throw new Error('Streaming not supported');

      const reader  = res.body.getReader();
      const decoder = new TextDecoder();
      let   buffer  = '';

      while (true) {
        const { done, value } = await reader.read();
        if (done) break;
        buffer += decoder.decode(value, { stream: true });
        const lines = buffer.split('\n');
        buffer = lines.pop();
        for (const line of lines) {
          if (line.startsWith('data: ')) {
            const data = line.slice(6).trim();
            if (data === '[DONE]') return;
            try {
              yield JSON.parse(data);
            } catch {}
          }
        }
      }
    }
  }

  // ── Convenience factory ───────────────────────────────────────────
  Yuga.create = (options) => new Yuga(options);

  return Yuga;
}));
